keps/sig-storage/1698-generic-ephemeral-volumes/README.md

```diff
@@ -99,28 +39,14 @@ SIG Architecture for cross cutting KEPs).
 
 ## Release Signoff Checklist
 
-<!--
-**ACTION REQUIRED:** In order to merge code into a release, there must be an
-issue in [kubernetes/enhancements] referencing this KEP and targeting a release
-milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
-of the targeted release**.
-
-For enhancements that make changes to code or processes/procedures in core
-Kubernetes i.e., [kubernetes/kubernetes], we require the following Release
-Signoff checklist to be completed.
-
-Check these off as they are completed for the Release Team to track. These
-checklist items _must_ be updated for the enhancement to be released.
--->
-
 Items marked with (R) are required *prior to targeting to a milestone / release*.
 
 - [X] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
-- [ ] (R) KEP approvers have approved the KEP status as `implementable`
-- [ ] (R) Design details are appropriately documented
-- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [X] (R) KEP approvers have approved the KEP status as `implementable`
+- [X] (R) Design details are appropriately documented
+- [X] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
 - [X] (R) Graduation criteria is in place
-- [ ] (R) Production readiness review completed
+- [X] (R) Production readiness review completed
```

Comment:
Again, sorry for commenting here, but I can't comment on lines you didn't change :(

<non PRR-related>
L226: Sorry for bringing this up if you already discussed that and I didn't find that. But I looked into PR and I didn't see security folks, so I bring this up (though I'm probably wrong here).

Allowing people to implicitly do things they don't have permissions for seems suspicious to me. Also note that once feature graduates to GA, we're locking down its feature gate to "always enabled" and remove it couple releases after, so operators won't really have ability to switch it off...
Given I don't feel competent to assess if this is acceptable or not, @liggitt - can you maybe help with accessing this?

Reply:
From an authz perspective, elevating pod create permission to allow triggering creation of N PVCs is surprising. Because this is proposing modifying the pod spec, it expands the surface area of every workload object, where only statefulsets could previously drive creation of PVC objects (and since that was part of the initial design, wasn't a surprise to people who had previously granted statefulset create permissions).

From a scalability perspective, did you look at how different components perform with the added write multiplication of a pod create triggering N PVC creations, and a pod delete triggering N PVC deletions? For example, can a deployment define a pod template that uses a few ephemeral volumes, scale up to 100-1000 pods, then roll out a new version?

Reply:
Security is indeed a concern. The ability to add ephemeral volumes through any app controller is what makes this approach appealing, but also the reason to be worried about security.

I had discussed this with @liggitt on Slack a while back before writing the KEP, just to get some feeling whether such a solution would be acceptable (no full review). At that time it boiled down to "if SIG-Storage thinks that this is needed, then let's consider it". Since then, the discussions around this KEP in SIG-Storage have confirmed that this is going to be useful for a variety of use cases that cannot be handled without this extension, so we would like to get it included.

> so operators won't really have ability to switch it off...

This is supposed to be addressed through the pod security policy: operators can enable a policy which allows some volume types, but not the new ephemeral inline volumes. In an earlier draft of the KEP I also had proposed to make it configurable per namespace via a label on the namespace, but @jsafrane then suggested the pod security policy approach and that seemed more elegant.

> From a scalability perspective, did you look at how different components perform with the added write multiplication of a pod create triggering N PVC creations, and a pod delete triggering N PVC deletions?

Not yet, and not even for "normal" provisioning through CSI. We started discussing scalability testing a year ago, but then didn't get around to actually doing anything, partly because observability was insufficient (no information about what is causing delays while creating or deleting volumes). The newish metrics support in the CSI sidecars might help with that now, and the upcoming distributed tracing feature should also be useful.

Ultimately, ephemeral inline volumes will perform as well (or badly) as that normal support. The only difference is that there will be more of those operations. Whether that is a problem depends largely on the kind of applications that will use them. For some use cases (like running memcached) I am not worried. I don't see that as something that needs to be run at such massive scale and even if it is needed, it's probably not going to be restarted that often. But we need to gather feedback from actual users here, which this alpha feature is supposed to enable.

Reply:
> The ability to add ephemeral volumes through any app controller is what makes this approach appealing

... and also that the implementation is very simple. I was able to get it working last week, without working on it full-time. Any other solution, like adding more features to the existing CSI ephemeral inline volume support, would be more complex and thus risky.

My intention was to submit a PR with the API change to start the API review as soon as possible (as mentioned here, we need clarify security concerns, but also whether embedding a `PersistentVolumeClaimSpec` may be sufficient), then continue with the implementation and adding tests (E2E tests exist, some unit tests are still missing).

Reply:
API PR here: https://github.com/kubernetes/kubernetes/pull/91680

Reply:
Security: @liggitt do you think the proposed PSP additions can address the concern? The PVCs that are created also will continue to go through storageclass quota.

Scalability: We did implement some scalability tests for volumes according to [the scale testing kep](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190204-scale-testing.md). They are available at https://github.com/kubernetes/perf-tests/tree/master/clusterloader2/testing/experimental/storage/pod-startup and currently support built-in ephemeral types like secrets and configmaps, and PVCs. Once we have an implementation, we can potentially run these tests with the CSI mock driver. I'm not sure we'll be able to match the performance of built-in ephemeral types, but we should understand and document the limits.

Reply:
I'm jumping into this late, sorry.  I don't know how common it is to restrict RBAC access to PVCs rather than restricting quota, but since it is expressible, it really does seem that we must respect that.  I don't know enough about the auth APIs, I admin, but it seems like every workload controller would need to check "can I" on PVCs for the original user, before doing that.  Is that viable?

Obviously, if you build or run a 3rd party controller that doesn't do this AND you give that controller the ability to make PVCs, you're on your own.  But we can DTRT in core and publish how...

Reply:
> and currently support built-in ephemeral types like secrets and configmaps, and PVCs

@msau42 PVCs are implemented, but not enabled because they were causing some issues on GCE storage side. xref:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/testing/experimental/storage/pod-startup/suite.yaml#L25
Are we ready to reenable them again?

Reply:
PVC using default storageclass probably still has gcp issues, but I think we should look into enabling the PVC tests using CSI mock driver, which can give us scale results of just the Kubernetes and CSI pieces without the storage backend.

Reply:
Going back to this thread, as @pohly pinged me.

Have the above concerns been addressed? I don't see any changes on this KEP, neither any changes here.
I think those has to be somehow addressed first (or I would like to see @liggitt to explicitly ACK on those changes from security POV - I don't feel competent and I think we need higher-level agreement before digging into PRR part.

Reply:
> Have the above concerns been addressed?

Yes. We went through this when reviewing the proposed implementation and then merged it into 1.19. You can consider the technical side of this KEP reviewed. What's missing is the PRR part for alpha (this PR) and later for beta (no PR yet - I intent to update the KEP accordingly once this PR is merged).

Alternatively, I can also use this PR to update the PRR part for beta in 1.21.

Reply:
Here's where Clayton and Tim approved the concept (including the security aspect):
-  https://github.com/kubernetes/kubernetes/pull/91680#issuecomment-656356694
- https://github.com/kubernetes/kubernetes/pull/92784#pullrequestreview-446027194

Reply:
@thockin, @smarterclayton : can you comment?

As you might remember, the key question for this feature was "how can we control who can use generic ephemeral inline volumes" and "do volume quotas still apply".

The solution for "who can use it" can't be RBAC rules because actual PVC creation is done by kube-controller-manager, not the user who created the pod. We settled on PSP because there are also other volume types that get controlled that way, so there was a precedent and it is consistent (https://github.com/kubernetes/kubernetes/blob/074a5177209c3605daf11cfe6f477fddce0d55c0/pkg/security/podsecuritypolicy/util/util.go#L40-L74).

For quotas the answer is simpler: those are enforced per namespace, and that works the same way regardless of who creates the PVC.

Reply:
Looking through all the discussions linked in https://github.com/kubernetes/kubernetes/pull/97171, it seems like sig-auth is still discussing possible replacements to PSP. Given that, I think we need to add a followup item for later to make sure we align with whatever replacement they decide on.

For now, assume we only have storageclass quota which can be enforced on the namespace level. Is that sufficient? Today you can restrict use of inline and pvcs at the pod and user level. So once this feature is enabled, users that were previously restricted can now use it if that namespace has storageclass quota. Is there a way we can close that gap?

Reply:
> Today you can restrict use of inline and pvcs at the pod and user level.

Which mechanism are you referring to here? PSP or RBAC?

RBAC controls which user can create PVCs. It does not control who can use those PVCs or other inline volumes inside a pod. As far as I know, that's only possible through PSP, so removal of PSP will create a gap that goes beyond this feature here. We just add one more thing (indirect creation of PVCs) that it is used for.

Reply:
I discussed with @liggitt on this. He suggested that to continue enforcing existing PSP policies that disallow pods using pvcs, we update psp enforcement to treat these new ephemeral volumes as a pvc type, instead of adding new fields to PSP.

Reply:
So a reference to a PVC and an inline PVC template shall be treated as the same thing as far as the PSP is concerned? Works for me.

I'll look at the implementation and update the KEP accordingly.

Reply:
> He suggested that to continue enforcing existing PSP policies that disallow pods using pvcs, we update psp enforcement to treat these new ephemeral volumes as a pvc type, instead of adding new fields to PSP.

I think something got lost in translation, sorry. I was saying the PSP enforcer should be updated to recognize these volume types, so an existing PSP that restricted pods to specific volume types would not suddenly start allowing this type.

I wasn't suggesting this type be treated as a PSP volume.

Reply:
I may be confused now :-) Is that different than saying "a policy that disallows PVCs will also disallow the pvc template"?

Reply:
> Is that different than saying "a policy that disallows PVCs will also disallow the pvc template"?

yes, it implies that a PSP that _allows_ PVC volumes would also allow this type.

A PSP with `volumes: ["secret","configmap","pvc"]` should allow PVCs and not allow this

A PSP with `volumes: ["secret","configmap"] should not allow PVCs or this

Reply:
Hm that means in order to use this feature you have to disable PSP altogether?

Reply:
> I was saying the PSP enforcer should be updated to recognize these volume types, so an existing PSP that restricted pods to specific volume types would not suddenly start allowing this type.

Isn't that exactly what was implemented already?

https://github.com/kubernetes/kubernetes/commit/c05c8e915b0c30d574405ed63e48a52801a1f75b added the new volume type to `pkg/security/podsecuritypolicy/util/util.go`. A negative test (= pod rejected) with the "default policy" (= no volume types listed) is here: https://github.com/kubernetes/kubernetes/blob/1f6fe21425a3963ee1549224a44fc48df5abbaed/pkg/security/podsecuritypolicy/provider_test.go#L500-L504

For the pod to be accepted, `psp.Spec.Volumes` must contain the type or "all": https://github.com/kubernetes/kubernetes/blob/1f6fe21425a3963ee1549224a44fc48df5abbaed/pkg/security/podsecuritypolicy/provider_test.go#L1027-L1041

There's no test for "pod rejected when some other volume type is listed", but that also works - I added a test locally and can include it in a future PR.

> Hm that means in order to use this feature you have to disable PSP for that pod?

If there is a PSP, it must be updated to allow generic ephemeral volumes. Without a PSP it works.

Reply:
commented at https://github.com/kubernetes/kubernetes/commit/c05c8e915b0c30d574405ed63e48a52801a1f75b#r46791570, hoisting here in case the commit gets lost/replaced:

adding a value to GetAllFSTypesAsSet expands the accepted values in the volumes PSP field. That's an API change that relaxes validation, so it has to be introduced in a ratcheting way over two releases (typically using the feature gate enablement):

* When the feature is disabled, tolerate this new value in updates of existing PSP objects that already contain the value, but do not allow it when creating a new PSP or updating a PSP that does not already contain the value.
* When the feature is enabled, allow this value on any create or update.

Reply:
That expansion of allowed values in the PSP field was what I was unsure of, since PSP is marked as deprecated

Reply:
That commit was already merged into Kubernetes for 1.19. I suppose it slipped through the review at that time.

So does that mean that for making this beta in 1.21, we need to add a feature gate check?

If we do that, do we have your okay to use PSP for controlling who may use generic ephemeral inline volumes?

Reply:
> That commit was already merged into Kubernetes for 1.19. I suppose it slipped through the review at that time.

Let me clarify: we discussed at length that we were extending PSP, so that part wasn't a surprise.

What slipped through is the part about having to add a feature gate check in that part of the code.

Reply:
I've extended the KEP with more information on how the PSP change needs to be handled - it's basically a copy of Jordan's comment.

@liggitt : can we get an explicit approval for this KEP PR from you? I'm also available for a short chat anytime today, if that helps.